build-tools/packages/build-cli/src/library/typescriptApi.ts

```diff
@@ -99,5 +117,73 @@ export function getApiExports(sourceFile: SourceFile): ExportRecords {
 		}
 	}
 
+	// If we have found levels for all things exported, then nothing else left to look for.
+	if (records.unknown.size === 0) {
+		return records;
+	}
+
+	// Otherwise, look for some special cases where level tagging may appear in source itself.
+
+	// Note that these are not exported declarations, but specifically
+	// given file's export declarations.
+	const exportDeclarations = sourceFile.getExportDeclarations();
+	for (const exportDeclaration of exportDeclarations) {
+		// console.log(`export @ ${exportDeclaration.getStartLineNumber()} has children:\n\t${exportDeclaration.getChildren().map((c) => c.getKindName()).join('\n\t')}`);
+
+		// Look first for namespace exports like `export * as foo`.
+		for (const namespaceDecl of exportDeclaration.getChildrenOfKind(
+			SyntaxKind.NamespaceExport,
+		)) {
+			const name = namespaceDecl.getName();
+			const unknownExported = records.unknown.get(name);
+			if (unknownExported !== undefined) {
+				console.log(
+					`namespace exports of the form 'export * as foo' are speculatively supported. See ${sourceFile.getFilePath()}:${namespaceDecl.getStartLineNumber()}:\n${namespaceDecl.getText()}`,
+				);
+				const namespaceLevel = getNodeLevel(exportDeclaration);
+				if (namespaceLevel === undefined) {
+					unknownExported.exportDecl = exportDeclaration;
+				} else {
+					records[namespaceLevel].push({ name, isTypeOnly: exportDeclaration.isTypeOnly() });
+					records.unknown.delete(name);
+					if (records.unknown.size === 0) {
+						return records;
+					}
+				}
+			}
+		}
+
+		// Then process named exports. Common case for named export is patch for api-extractor
+		// limitation on namespace exports:
+		//   import * as foo from "bar";
+		//   export { foo }
+		// Entire foo namespace may be tagged by:
+		//   export {
+		//     /** @internal */
+		//     foo
+		//   }
+		// Note that per namespace level tags maybe different than the namespace. To avoid any
```

Comment:
I suspect this is why API-Extractor hasn't implemented this yet. In theory, we need to recurse into the namespace's child items and do the appropriate trimming at that layer. I think namespaces are a case where redeclaration is necessary, and I think it's safe to do so. 

I wonder if we could at least add some validation here that there aren't any items being included that shouldn't. How easy would this be?

Reply:
Presumable you should be able to enumerate all of the things inside the namespace/object being exported.
It would probably be better to have multiple namespaces. Redeclaring is where we get into trouble. These files need to export exactly what is in the original because that is what will be there at runtime.
It would be okay to do something like type manipulation to Omit members - don't know if such a thing is possible.